java/rocksjni/portal.h

```diff
@@ -9089,5 +9089,155 @@ class BlockBasedTableOptionsJni
   }
 };
 
+class WideColumnJni {
+ public:
+  static void convertWideColumnsDirect(
+      std::function<void(ROCKSDB_NAMESPACE::Slice&,
+                         ROCKSDB_NAMESPACE::WideColumns&)>
+          op,
+      JNIEnv* env, jobject jKey, jint jKeyOffset, jint jKeyLen,
+      jobjectArray jNames, jintArray jNamesOffset, jintArray jNamesLen,
+      jobjectArray jValues, jintArray jvaluesOffset, jintArray jValuesLen) {
+    auto _key = env->GetDirectBufferAddress(jKey);
+    if (_key == nullptr) {
+      ROCKSDB_NAMESPACE::RocksDBExceptionJni::ThrowNew(
+          env,
+          "Invalid key argument (argument is not a valid direct ByteBuffer)");
+      return;
+    }
+
+    auto key = reinterpret_cast<char*>(_key) + jKeyOffset;
+    ROCKSDB_NAMESPACE::Slice key_slice(key, static_cast<size_t>(jKeyLen));
+
+    const auto columns_len = static_cast<int>(env->GetArrayLength(jNames));
+
+    auto namesOffset = std::make_unique<jint[]>(columns_len);
+    env->GetIntArrayRegion(jNamesOffset, 0, columns_len, namesOffset.get());
+    if (env->ExceptionCheck()) {
+      return;
+    }
+    auto namesLength = std::make_unique<jint[]>(columns_len);
+    env->GetIntArrayRegion(jNamesLen, 0, columns_len, namesLength.get());
+    if (env->ExceptionCheck()) {
+      return;
+    }
+
+    auto valuesOffset = std::make_unique<jint[]>(columns_len);
+    env->GetIntArrayRegion(jvaluesOffset, 0, columns_len, valuesOffset.get());
+    if (env->ExceptionCheck()) {
+      return;
+    }
+
+    auto valuesLenght = std::make_unique<jint[]>(columns_len);
+    env->GetIntArrayRegion(jValuesLen, 0, columns_len, valuesLenght.get());
+    if (env->ExceptionCheck()) {
+      return;
+    }
+
+    ROCKSDB_NAMESPACE::WideColumns columns;
+    columns.reserve(columns_len);
+
+    for (int i = 0; i < columns_len; i++) {
+      auto jname = env->GetObjectArrayElement(jNames, i);
+      if (env->ExceptionCheck()) {
+        return;
+      }
+      auto _name = env->GetDirectBufferAddress(jname);
+      if (_name == nullptr) {
+        ROCKSDB_NAMESPACE::RocksDBExceptionJni::ThrowNew(
+            env,
+            "Invalid \"name\" argument (argument is not a valid direct "
+            "ByteBuffer)");
+        return;
+      }
+      auto name = reinterpret_cast<char*>(_name) + namesOffset[i];
```

Comment:
There are utility methods in portal.h to go from `jstring` to `std::string` or `char*` - could we use those here instead?

Reply:
We are not working with `jstring` but with `void*` here. I cast it to `char*` so that I can add offset to it.